backends/magma/ceed-magma-basis.c

```diff
@@ -257,401 +254,441 @@ CEED_INTERN "C"
-
-  CeedInt Narray[MAGMA_NONTENSOR_KERNEL_INSTANCES] = {MAGMA_NONTENSOR_N_VALUES};
-  CeedInt iN                                       = 0;
-  CeedInt diff                                     = abs(Narray[iN] - N);
-  for (CeedInt in = iN + 1; in < MAGMA_NONTENSOR_KERNEL_INSTANCES; in++) {
-    CeedInt idiff = abs(Narray[in] - N);
-    if (idiff < diff) {
-      iN   = in;
-      diff = idiff;
+  CeedInt P = ndof, Q = nqpts, N = nelem * ncomp, NB = 1, iN = -1;
+  CeedInt M = (tmode == CEED_TRANSPOSE) ? P : Q, K = (tmode == CEED_TRANSPOSE) ? Q : P;
+  CeedInt Narray[MAGMA_NONTENSOR_KERNEL_INSTANCES] = {MAGMA_NONTENSOR_KERNEL_N_VALUES};
+  if (P <= MAGMA_NONTENSOR_CUSTOM_KERNEL_MAX_P && Q <= MAGMA_NONTENSOR_CUSTOM_KERNEL_MAX_Q) {
+    iN           = 0;
+    CeedInt diff = abs(Narray[iN] - N);
+    for (CeedInt in = iN + 1; in < MAGMA_NONTENSOR_KERNEL_INSTANCES; in++) {
+      CeedInt idiff = abs(Narray[in] - N);
+      if (idiff < diff) {
+        iN   = in;
+        diff = idiff;
+      }
     }
+    NB = nontensor_rtc_get_nb(magma_getdevice_arch(), 'd', tmode, qcomp, P, Narray[iN], Q);
   }
 
-  NB     = nontensor_rtc_get_nb(arch, 'd', emode, tmode, P, Narray[iN], Q);
-  interp = (tmode == CEED_TRANSPOSE) ? &impl->magma_interp_tr_nontensor[iN] : &impl->magma_interp_nontensor[iN];
-  grad   = (tmode == CEED_TRANSPOSE) ? &impl->magma_grad_tr_nontensor[iN] : &impl->magma_grad_nontensor[iN];
-
-  switch (emode) {
-    case CEED_EVAL_INTERP: {
-      CeedInt P = ndof, Q = nqpt;
-      if (P < MAGMA_NONTENSOR_CUSTOM_KERNEL_MAX_P && Q < MAGMA_NONTENSOR_CUSTOM_KERNEL_MAX_Q) {
-        CeedInt M     = (tmode == CEED_TRANSPOSE) ? P : Q;
-        CeedInt K     = (tmode == CEED_TRANSPOSE) ? Q : P;
-        CeedInt ntcol = MAGMA_NONTENSOR_BASIS_NTCOL(M);
-        CeedInt shmem = 0, shmemA = 0, shmemB = 0;
-        shmemB += ntcol * K * NB * sizeof(CeedScalar);
-        shmemA += (tmode == CEED_TRANSPOSE) ? 0 : K * M * sizeof(CeedScalar);
-        shmem = (tmode == CEED_TRANSPOSE) ? (shmemA + shmemB) : CeedIntMax(shmemA, shmemB);
-
-        CeedInt       grid   = MAGMA_CEILDIV(MAGMA_CEILDIV(N, NB), ntcol);
-        magma_trans_t transA = (tmode == CEED_TRANSPOSE) ? MagmaNoTrans : MagmaTrans;
-        magma_trans_t transB = MagmaNoTrans;
-        CeedScalar    alpha = 1.0, beta = 0.0;
-
-        void *args[] = {&transA, &transB, &N, &alpha, &impl->dinterp, &P, &du, &K, &beta, &dv, &M};
-        CeedCallBackend(CeedRunKernelDimSharedMagma(ceed, *interp, grid, M, ntcol, 1, shmem, args));
-      } else {
-        if (tmode == CEED_TRANSPOSE)
-          magma_gemm_nontensor(MagmaNoTrans, MagmaNoTrans, P, nelem * ncomp, Q, 1.0, impl->dinterp, P, du, Q, 0.0, dv, P, data->queue);
-        else magma_gemm_nontensor(MagmaTrans, MagmaNoTrans, Q, nelem * ncomp, P, 1.0, impl->dinterp, P, du, P, 0.0, dv, Q, data->queue);
+  // Compile kernels for NB as needed (blocking size for N and NB from first call is used for all future calls)
+  if (!impl->module) {
```

Comment:
Hi @sebastiangrimberg, thanks for all your work on this.  I haven't looked through in detail yet, but I wanted to go ahead and comment on this part.  I don't think this behavior is what we want.  I agree that it is likely to not matter in a lot of practical use cases.  But technically, according to the libCEED API, a user could create one basis object to use with two very different operators -- e.g. what if one of them is only tied to a specific part of the local mesh, a region of interest where some quantity is evaluated, and the two operators therefore have different values of $N$ -- different enough that we would want to choose different tuning parameters for them?  Again, I'm not sure how common such a use case would be.  But I think we have to respect that the API allows for it.  "Locking in" a basis object's tuning with its first application could cause some weird/unexpected results for users (why is the performance of a similar operator so much better in this one application vs when I use it over here in this application?).  I suppose we could add documentation specifying that if you have such a use case in mind, you should create two separate basis objects to use, even if they are identical in basis type/size/dimension.  But that feels rather clunky to me, and like something we shouldn't force the user to take care of for us.

If we're concerned about compile time during basis creation, we could only compile each version of the kernel as needed by checking during the basis apply.  I know that's more complicated, but I think it's potentially important.  The QFunction already delays compilation until its first application (though of course it only ever needs to be compiled once, and wouldn't need to store multiple kernels).

Reply:
Yeah I was worried if this might be a use case we want to support. In that case we can still do the lazy compilation and we will just allow for caching the different compiled kernels for each value of `N` instead of just one.

Reply:
Hi @nbeams, the latest commit https://github.com/CEED/libCEED/pull/1301/commits/60805e1fa49636f8be12a4d108cd17c389a19df5 should resolve this. Basically we now allow caching of JiT kernels for each possible value of `N` and check if one exists during `CeedBasisApply`, else we RTC it. Let me know what you think and if this resolves the concern.

Note that the weight kernels are independent of `N` so we can just compile them once for all during basis construction, which is nice. But this requires storing a separate `CUmodule/hipModule_t`. I think that's fine though.

Reply:
Thanks! Have you tested this locally? I don't know if we have any tests that would capture the full behavior (not just the delay in compilation, but if we end up needing to compile 2 or more different instances throughout the lifetime of the basis object).

Reply:
I have not tested this locally, as I'm in the same case of not having any tests for this type of behavior (hence my removing it on the first go around 😄 ).